libnmstate/state.py

```diff
@@ -0,0 +1,248 @@
+        return self._ifaces_state
+
+    def sanitize_ethernet(self, other_state):
+        """
+        Given the other_state, update the ethernet interfaces state base on
+        the other_state ethernet interfaces data.
+        Usually the other_state represents the current state.
+        If auto-negotiation, speed and duplex settings are not provided,
+        but exist in the current state, they need to be set to None
+        to not override them with the values from the current settings
+        since the current settings are read from the device state and not
+        from the actual configuration.  This makes it possible to distinguish
+        whether a user specified these values in the later configuration step.
+        """
+        for ifname, iface_state in six.viewitems(self.interfaces):
+            iface_current_state = other_state.interfaces.get(ifname, {})
+            if iface_current_state.get('type') == 'ethernet':
+                ethernet = iface_state.setdefault('ethernet', {})
+                ethernet.setdefault('auto-negotiation', None)
+                ethernet.setdefault('speed', None)
+                ethernet.setdefault('duplex', None)
+
+    def sanitize_dynamic_ip(self):
+        """
+        If dynamic IP is enabled and IP address is missing, set an empty
+        address list. This assures that the desired state is not complemented
+        by the current state address values.
+        If dynamic IP is disabled, all dynamic IP options should be removed.
+        """
+        for iface_state in six.viewvalues(self.interfaces):
+            for family in ('ipv4', 'ipv6'):
+                ip = iface_state.get(family, {})
+                if ip.get('enabled') and (
+                        ip.get('dhcp') or ip.get('autoconf')):
+                    ip['address'] = []
+                else:
+                    for dhcp_option in ('auto-routes',
+                                        'auto-gateway',
+                                        'auto-dns'):
+                        ip.pop(dhcp_option, None)
+
+    def verify_interfaces(self, other_state):
+        """Verify that the (self) state is a subset of the other_state. """
+        self._remove_absent_interfaces()
+        self._remove_down_virt_interfaces()
+
+        self._assert_interfaces_included_in(other_state)
+
+        metadata.remove_ifaces_metadata(self)
+        other_state.sanitize_dynamic_ip()
+
+        self.canonicalize_interfaces(other_state)
+
+        self.normalize_for_verification()
+        other_state.normalize_for_verification()
+
+        self._assert_interfaces_equal(other_state)
+
+    def normalize_for_verification(self):
+        self._clean_sanitize_ethernet()
+        self._sort_lag_slaves()
+        self._sort_bridge_ports()
+        self._canonicalize_ipv6()
+        self._remove_iface_ipv6_link_local_addr()
+        self._sort_ip_addresses()
+
+    def canonicalize_interfaces(self, other_state):
```

Comment:
IMHO this should be called `merge()` since canonicalization also means normalization.

Reply:
Sounds reasonable to me, but you need to add `interfaces` at the moment (perhaps this will change).
Lets do it in a separate small PR, I prefer to keep this as the original name so it can be tracked.